pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -213,17 +210,22 @@ type Controller struct {
 
 // NewController creates a new Kubernetes controller
 // Created by bootstrap and multicluster (see secretcontroler).
-func NewController(client kubernetes.Interface, metadataClient metadata.Interface, options Options) *Controller {
-	log.Infof("Service controller watching namespace %q for services, endpoints, nodes and pods, refresh %s",
-		options.WatchedNamespaces, options.ResyncPeriod)
-
-	watchedNamespaceList := strings.Split(options.WatchedNamespaces, ",")
+func NewController(
+	client kubernetes.Interface,
+	podInformer coreinformers.PodInformer,
+	nodeInformer coreinformers.NodeInformer,
+	serviceInformer coreinformers.ServiceInformer,
+	endpointsInformer coreinformers.EndpointsInformer,
+	endpointSliceInformer v1alpha1.EndpointSliceInformer,
+	nodeMetadataInformer informers.GenericInformer,
+	metadataClient metadata.Interface,
+	options Options) *Controller {
```

Comment:
yikes. lots of args. Why can't we turn the kubelib into a factory? consumers can do kubehandle.GetPodInformer(), kubehandle.GetClient, etc.. This way, k8s controller's args are just the kubehandle and options.

Reply:
I pretty much ripped off what k8s is doing: https://github.com/kubernetes/kubernetes/blob/c6fd25d10032b22ca28e437957390f89a4476ecf/pkg/controller/volume/attachdetach/attach_detach_controller.go#L105.
I am fine with passing the whole kube.Client in though